core/templates/expressions/expression-syntax-tree.service.ts

```diff
@@ -16,417 +16,447 @@
-            paramsFound.push(parseTree[1]);
-          } else {
-            for (var i = 1; i < parseTree.length; i++) {
-              paramsFound = paramsFound.concat(_findParams(parseTree[i]));
-            }
+  public getParamsUsedInExpression(expression: string): Array<string> {
+    let _findParams = (parseTree: Array<string>|string): Array<string> => {
+      let paramsFound = [];
+      if (parseTree instanceof Array) {
+        if (parseTree[0] === '#') {
+          paramsFound.push(parseTree[1]);
+        } else {
+          for (let i = 1; i < parseTree.length; i++) {
+            paramsFound = paramsFound.concat(_findParams(parseTree[i]));
           }
         }
+      }
 
-        var uniqueParams = [];
-        for (var i = 0; i < paramsFound.length; i++) {
-          if (uniqueParams.indexOf(paramsFound[i]) === -1) {
-            uniqueParams.push(paramsFound[i]);
-          }
+      let uniqueParams = [];
+      for (let i = 0; i < paramsFound.length; i++) {
+        if (uniqueParams.indexOf(paramsFound[i]) === -1) {
+          uniqueParams.push(paramsFound[i]);
         }
+      }
 
-        return uniqueParams.sort();
-      };
-
-      var parsed = ExpressionParserService.parse(expression);
-      return _findParams(parsed);
+      return uniqueParams.sort();
     };
 
-    // Checks if the args array has the expectedNum number of elements and
-    // throws an error if not. If optional expectedMax is specified, it
-    // verifies the number of args is in [expectedNum, expectedMax] range
-    // inclusive.
-    var verifyNumArgs = function(args, expectedNum, expectedMax = expectedNum) {
-      if (expectedMax === undefined) {
-        expectedMax = expectedNum;
-      }
-      if (args.length >= expectedNum && args.length <= expectedMax) {
-        return;
-      }
-      throw new ExprWrongNumArgsError(args, expectedNum, expectedMax);
-    };
+    let parsed = this.expressionParserService.parse(expression);
 
-    var _verifyArgTypesMatchExpectedType = function(argTypes, expectedType) {
-      for (var i = 0; i < argTypes.length; i++) {
-        if (argTypes[i] !== expectedType) {
-          throw new ExprWrongArgTypeError(null, argTypes[i], expectedType);
-        }
-      }
-      return true;
-    };
+    return _findParams(parsed);
+  }
 
-    var _verifyArgTypesMatch = function(argType1, argType2) {
-      if (argType1 !== argType2) {
-        throw new ExprWrongArgTypeError(null, argType1, argType2);
+  // Checks if the args array has the expectedNum number of elements and
+  // throws an error if not. If optional expectedMax is specified, it
+  // verifies the number of args is in [expectedNum, expectedMax] range
+  // inclusive.
+  private verifyNumArgs(args: Array<string>, expectedNum: number,
+      expectedMax: number = expectedNum): void {
+    if (expectedMax === undefined) {
+      expectedMax = expectedNum;
+    }
+    if (args.length >= expectedNum && args.length <= expectedMax) {
```

Comment:
Please invert this condition and throw from it instead.

Reply:
Careful, the opposite of (A && B) is (!A || !B) (see: [De Morgan's laws](https://en.wikipedia.org/wiki/De_Morgan%27s_laws), and the opposite of `a <= b` is `a > b` so you should change it to:

```typescript
if (args.length < expectedNum || args.length > expectedMax) {
      throw new ExprWrongNumArgsError(args, expectedNum, expectedMax);
}
```

Reply:
@brianrodri I must admit, I something just learned. Thanks!

Reply:
Happy I could help you learn something new :)